### PR TITLE
[mlir] notify insertion of parent op first when cloning

### PR DESCRIPTION
**Pull request by jeanPerier**
When cloning an operation with a region, the builder was currently notifying about the insertion of the cloned operations inside the region before the cloned operation itself.

When using cloning inside rewrite pass, this could cause issues if a pattern is expected to be applied on a cloned parent operation before trying to apply patterns on the cloned operations it contains (the patterns are attempted in order of notifications for the cloned operations).

### CODE REVIEW

**Review by tblah — COMMENTED** [review]
Changes LGTM, but please wait for approval from someone with more expertise in MLIR.

**Review by joker-eph — COMMENTED** [review]
I'm wary of patterns being sensitive to this kind of details, but the change itself LGTM otherwise.

(Nice finding for the missing FileCheck!)

File: mlir/lib/IR/Builders.cpp
Comment (matthias-springer) [inline mlir/lib/IR/Builders.cpp:531]: The `insert` call is now above.

File: mlir/lib/IR/Builders.cpp
Comment (matthias-springer) [inline mlir/lib/IR/Builders.cpp:539]: Your change is changing the notification order for `newOp` from post-order to pre-order. For consistency, `region.walk` should also be a pre-order walk now.

**Comment by matthias-springer** [comment]
This change makes sense to me: when creating IR with the builder API, outer operations are typically created before inner operations. So it makes sense to notify in that order.

(But patterns should not depend on a particular application order (apart from `benefit`).)

File: mlir/lib/IR/Builders.cpp
Comment (matthias-springer) [reply mlir/lib/IR/Builders.cpp:530]: `pattern converter` Do you mean dialect conversion?

Comment (vzakhari) [reply mlir/lib/IR/Builders.cpp:530]: Yes, I meant the dialect conversion.

File: mlir/lib/IR/Builders.cpp
Comment (jeanPerier) [reply mlir/lib/IR/Builders.cpp:531]: Fixed.

File: mlir/lib/IR/Builders.cpp
Comment (jeanPerier) [reply mlir/lib/IR/Builders.cpp:530]: No, the conversion is not attempted right when being notified about the insertion. When cloning inside a pattern, the legalization for the cloned operation is only done when the replacement/root update is finalized. The legalization is done in the same order the notification were emitted.
This is just like when creating an operation with a region inside a pattern directly (without cloning).

File: mlir/lib/IR/Builders.cpp
Comment (jeanPerier) [reply mlir/lib/IR/Builders.cpp:539]: Thanks for catching this! Updated the walk to use pre-order and the test to verify that.

**Review by vzakhari — APPROVED** [review]
LGTM

**Review by matthias-springer — APPROVED** [review]

**Pull request merged**
